SKL 1.4 release breaks build

Scikit-learn 1.4.0 was [released](https://github.com/scikit-learn/scikit-learn/releases/tag/1.4.0-1) today.

We have the following errors ([ci run](https://github.com/microsoft/hummingbird/actions/runs/7589300833/job/20673644648)):

```
 =========================== short test summary info ============================
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_linear_regression - TypeError: BaggingRegressor.__init__() got an unexpected keyword argument 'base_estimator'
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_linear_svr - TypeError: BaggingRegressor.__init__() got an unexpected keyword argument 'base_estimator'
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_linear_svr_1 - TypeError: BaggingRegressor.__init__() got an unexpected keyword argument 'base_estimator'
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_logistic_regression - TypeError: BaggingClassifier.__init__() got an unexpected keyword argument 'base_estimator'
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_svc - TypeError: BaggingClassifier.__init__() got an unexpected keyword argument 'base_estimator'
FAILED tests/test_sklearn_bagging.py::TestSklearnBagging::test_bagging_svc_1 - TypeError: BaggingClassifier.__init__() got an unexpected keyword argument 'base_estimator'
===== 6 failed, 359 passed, 275 skipped, 69 warnings in 125.83s (0:02:05) =================
```

And also [failures in](https://github.com/microsoft/hummingbird/actions/runs/7589779315/job/20675116672):
`tests\test_extra_conf.py ........FF.Fs............ssssssssssss`

```
>           raise MissingConverter("Unable to find converter for model type {}.".format(model_type))
E           hummingbird.ml.exceptions.MissingConverter: Unable to find converter for model type <class 'sklearn.preprocessing._function_transformer.FunctionTransformer'>.
```
See https://github.com/scikit-learn/scikit-learn/pull/27204/

and

`tests\test_sklearn_pipeline.py ....F.FFFFFFF.FF.......`

```
>       categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
E       TypeError: OneHotEncoder.__init__() got an unexpected keyword argument 'sparse'
```
See https://github.com/scikit-learn/scikit-learn/pull/27861

--- a/hummingbird/ml/_parse.py
+++ b/hummingbird/ml/_parse.py
@@ -21,7 +21,7 @@ from sklearn.compose import ColumnTransformer
 from sklearn.ensemble import BaggingClassifier, BaggingRegressor
 from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
 from sklearn.multioutput import MultiOutputRegressor, RegressorChain
-from sklearn.preprocessing import OneHotEncoder
+from sklearn.preprocessing import OneHotEncoder, FunctionTransformer
 
 from .containers import CommonSklearnModelContainer, CommonONNXModelContainer, CommonSparkMLModelContainer
 from ._topology import Topology
@@ -35,7 +35,7 @@ try:
 except ImportError:
     StackingClassifier = None
 
-do_not_merge_columns = tuple(filter(lambda op: op is not None, [OneHotEncoder, ColumnTransformer]))
+do_not_merge_columns = tuple(filter(lambda op: op is not None, [OneHotEncoder, ColumnTransformer, FunctionTransformer]))
 
 
 def parse_sklearn_api_model(model, extra_config={}):
@@ -473,6 +473,11 @@ def _parse_sklearn_column_transformer(topology, model, inputs):
     return transformed_result_names
 
 
+def _parse_sklearn_function_transformer(topology, model, inputs):
+    # TODO borrow from _parse_sklearn_column_transformer
+    pass
+
+
 def _parse_sklearn_stacking(topology, model, inputs):
     """
     Taken from https://github.com/onnx/sklearn-onnx/blob/9939c089a467676f4ffe9f3cb91098c4841f89d8/skl2onnx/_parse.py#L238.
@@ -549,6 +554,7 @@ def _build_sklearn_api_parsers_map():
     # Parsers for edge cases are going here.
     map_parser = {
         ColumnTransformer: _parse_sklearn_column_transformer,
+        FunctionTransformer: _parse_sklearn_function_transformer,
         GridSearchCV: _parse_sklearn_model_selection,
         MultiOutputRegressor: _parse_sklearn_multi_output_regressor,
         pipeline.Pipeline: _parse_sklearn_pipeline,

--- a/hummingbird/ml/_parse.py
+++ b/hummingbird/ml/_parse.py
@@ -474,8 +474,23 @@ def _parse_sklearn_column_transformer(topology, model, inputs):
 
 
 def _parse_sklearn_function_transformer(topology, model, inputs):
-    # TODO borrow from _parse_sklearn_column_transformer
-    pass
+    """
+    Taken from https://github.com/onnx/sklearn-onnx/blob/fdb52cec86d4d19401cc365db97650fd7692676b/skl2onnx/operator_converters/function_transformer.py#L10.  # noqa: E501
+    :param topology: Topology object
+    :param model: A *scikit-learn* *FunctionTransformer* object
+    :param inputs: A list of Variable objects
+    :return: Output produced by function transformer
+    """
+    if len(inputs) == 1:
+        # apply identity
+        return inputs
+    else:
+        # apply concat
+        conc_op = topology.declare_logical_operator("SklearnConcat")
+        conc_op.inputs = inputs
+        conc_names = topology.declare_logical_variable("concat_inputs")
+        conc_op.outputs.append(conc_names)
+        return conc_op.outputs
 
 
 def _parse_sklearn_stacking(topology, model, inputs):

--- a/tests/test_sklearn_bagging.py
+++ b/tests/test_sklearn_bagging.py
@@ -13,7 +13,7 @@ from hummingbird.ml import constants
 class TestSklearnBagging(unittest.TestCase):
     def test_bagging_svc_1(self):
         X, y = make_classification(n_samples=100, n_features=4, n_informative=2, n_redundant=0, random_state=0, shuffle=False)
-        clf = BaggingClassifier(base_estimator=SVC(), n_estimators=10, random_state=0)
+        clf = BaggingClassifier(estimator=SVC(), n_estimators=10, random_state=0)
 
         clf.fit(X, y)
 
@@ -25,7 +25,7 @@ class TestSklearnBagging(unittest.TestCase):
 
     def test_bagging_svc(self):
         X, y = make_classification(n_samples=100, n_features=4, n_informative=2, n_redundant=0, random_state=0, shuffle=False)
-        clf = BaggingClassifier(base_estimator=SVC(), n_estimators=10, random_state=0)
+        clf = BaggingClassifier(estimator=SVC(), n_estimators=10, random_state=0)
 
         clf.fit(X, y)
 
@@ -37,7 +37,7 @@ class TestSklearnBagging(unittest.TestCase):
 
     def test_bagging_logistic_regression(self):
         X, y = make_classification(n_samples=100, n_features=4, n_informative=2, n_redundant=0, random_state=0, shuffle=False)
-        clf = BaggingClassifier(base_estimator=LogisticRegression(), n_estimators=10, random_state=0)
+        clf = BaggingClassifier(estimator=LogisticRegression(), n_estimators=10, random_state=0)
 
         clf.fit(X, y)
 
@@ -49,7 +49,7 @@ class TestSklearnBagging(unittest.TestCase):
 
     def test_bagging_linear_svr_1(self):
         X, y = make_regression(n_samples=1000, n_features=8, n_informative=5, n_targets=1, random_state=0, shuffle=True)
-        reg = BaggingRegressor(base_estimator=LinearSVR(), n_estimators=10, random_state=0)
+        reg = BaggingRegressor(estimator=LinearSVR(), n_estimators=10, random_state=0)
 
         reg.fit(X, y)
 
@@ -64,7 +64,7 @@ class TestSklearnBagging(unittest.TestCase):
 
     def test_bagging_linear_svr(self):
         X, y = make_regression(n_samples=1000, n_features=8, n_informative=5, n_targets=1, random_state=0, shuffle=True)
-        reg = BaggingRegressor(base_estimator=LinearSVR(), n_estimators=10, random_state=0)
+        reg = BaggingRegressor(estimator=LinearSVR(), n_estimators=10, random_state=0)
 
         reg.fit(X, y)
 
@@ -76,7 +76,7 @@ class TestSklearnBagging(unittest.TestCase):
 
     def test_bagging_linear_regression(self):
         X, y = make_regression(n_samples=1000, n_features=8, n_informative=5, n_targets=1, random_state=0, shuffle=True)
-        reg = BaggingRegressor(base_estimator=LinearRegression(), n_estimators=10, random_state=0)
+        reg = BaggingRegressor(estimator=LinearRegression(), n_estimators=10, random_state=0)
 
         reg.fit(X, y)
 

--- a/tests/test_sklearn_pipeline.py
+++ b/tests/test_sklearn_pipeline.py
@@ -204,7 +204,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -245,7 +245,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -286,7 +286,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -327,7 +327,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -369,7 +369,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -411,7 +411,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -454,7 +454,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -496,7 +496,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -539,7 +539,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[
@@ -581,7 +581,7 @@ class TestSklearnPipeline(unittest.TestCase):
 
         numeric_transformer = Pipeline(steps=[("scaler", StandardScaler())])
 
-        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse=True, handle_unknown="ignore"))])
+        categorical_transformer = Pipeline(steps=[("onehot", OneHotEncoder(sparse_output=True, handle_unknown="ignore"))])
 
         preprocessor = ColumnTransformer(
             transformers=[

